misc/python/materialize/mzexplore/analyze.py

```diff
@@ -0,0 +1,93 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+"""
+Utilities to extract data from a Materialize catalog for exploration purposes.
+"""
+
+import textwrap
+from pathlib import Path
+from typing import TextIO
+
+from materialize.mzexplore.common import info, warn
+
+
+def changes(
+    out: TextIO,
+    target: Path,
+    header_name: str,
+    base_suffix: str,
+    diff_suffix: str,
+) -> None:
+    """
+    Append sections to an `*.md` file with items corresponding to changes in
+    plans.
+    """
+
+    # Ensure that the target dir exists
+    if not target.is_dir():
+        warn(f"Target path `{target}` is not a folder")
+        return
+
+    info(
+        f"Comparing `{base_suffix}` vs `{diff_suffix}` "
+        f"plans for environment {header_name}"
+    )
+
+    info("Comparing optimized plans")
+    out.write(
+        textwrap.dedent(
+            f"""
+            ## {header_name}
+
+            """
+        ).lstrip("\n")
+    )
+
+    for base in target.glob(f"**/*.optimized_plan.{base_suffix}.txt"):
+        name = base.name.split(".")
+        ext = name.pop()  # extension
+        name.pop()  # suffix
+        stage = name.pop()  # explain stage
+        typ = name.pop()  # explain type
+        item = ".".join(name)
+
+        parents = list(base.parents)
+        parents.reverse()
+        schema = parents.pop().name
+        database = parents.pop().name
+        item_type = parents.pop().name
+
+        diff = base.parent / f"{item}.{typ}.{stage}.{diff_suffix}.{ext}"
+
+        if not diff.exists():
```

Comment:
Warn? Or is this just what we'll get when there's no change?

Reply:
:+1: